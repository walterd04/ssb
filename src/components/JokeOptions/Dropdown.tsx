import React, { useCallback, useState, useEffect, useRef } from "react";
import styled from "styled-components";
import OutsideClickHandler from "react-outside-click-handler";

import caretDown from "../../assets/caret.svg";

type DropdownOption = { label: string; value: string };

interface DropdownProps {
  id: string;
  onClick?: () => void;
  onChangeEventHandler: (e: DropdownOption) => void;
  items: DropdownOption[];
  placeholder: string;
}

const HiddenItem = styled.div`
  display: none;
`;

const DropdownContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-start;
  background-color: #ffffff;
  border: 1px solid #000000;
  box-shadow: 3px 3px 0px rgba(0, 0, 0, 0.25);
`;

const DropdownSelect = styled.div`
  display: flex;
  flex-flow: row nowrap;
  position: relative;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  height: 48px;
  padding: 12px;
`;

const StyledDropdown = styled.div`
  display: block;
  position: relative;
  width: 90%;
  margin: 0 auto;
`;
const DropdownPlaceholder = styled.div`
  display: flex;
  flex-direction: row;
  width: 100%;
  align-items: center;
`;

const DropdownItem = styled.button`
  display: block;
  position: relative;
  padding: 12px 0;
  width: 100%;
  appearance: none;
  text-align: left;
`;

type CaretProps = { isActive: boolean };

const Caret = styled.img<CaretProps>`
  width: 10px;
  height: 10px;
  margin-left: auto;
  transform: ${(props) => (props.isActive ? "rotate(180deg)" : "")};
`;

export const Dropdown: React.FC<DropdownProps> = ({
  id,
  items,
  placeholder,
  onClick,
  onChangeEventHandler,
}: DropdownProps) => {
  const [active, setActive] = useState<boolean>(false);
  const [selectPlaceholder, setSelectPlaceholder] = useState<string>(
    placeholder
  );
  const [activeValue, setActiveValue] = useState<string | undefined>(undefined);

  const UIDropdown = useRef<HTMLDivElement | null>(null);

  const handleClick = useCallback((): void => {
    setActive(!active);
    if (onClick) {
      onClick();
    }
  }, [onClick, setActive, active]);

  useEffect(() => {
    const renderedDropdown: HTMLDivElement | null = UIDropdown.current;
    if (renderedDropdown)
      renderedDropdown.addEventListener("keypress", () => {
        handleClick();
      });
    return () => {
      if (!renderedDropdown) {
        return;
      }
      renderedDropdown.removeEventListener("keypress", handleClick, true);
    };
  }, [active, setActive, onClick, handleClick]);

  const handleOptionClick = (item: DropdownOption): void => {
    setActive(false);
    setSelectPlaceholder(item.label);
    setActiveValue(item.value);
    if (onChangeEventHandler) onChangeEventHandler(item);
  };

  return (
    <OutsideClickHandler onOutsideClick={() => setActive(false)}>
      <HiddenItem>
        <select
          data-testid="hidden-select"
          id={id}
          name={id}
          onChange={(e: unknown) => console.log(e)}
          value={activeValue}
        >
          {items.map((item: DropdownOption) => (
            <option key={item.value} value={item.value}>
              {item.label}
            </option>
          ))}
        </select>
      </HiddenItem>
      <DropdownContainer>
        <DropdownSelect
          data-testid="visual-select"
          ref={UIDropdown}
          onClick={handleClick}
          tabIndex={0}
          title={selectPlaceholder}
        >
          <DropdownPlaceholder>
            {selectPlaceholder}
            <Caret isActive={active} src={caretDown} alt="arrow down" />
          </DropdownPlaceholder>
        </DropdownSelect>
        {active && (
          <StyledDropdown data-testid="visual-dropdown">
            {items.map((item: { value: string; label: string }) => (
              <DropdownItem
                onClick={() => handleOptionClick(item)}
                key={item.value}
                data-testid={item.value}
              >
                {item.label}
              </DropdownItem>
            ))}
          </StyledDropdown>
        )}
      </DropdownContainer>
    </OutsideClickHandler>
  );
};
