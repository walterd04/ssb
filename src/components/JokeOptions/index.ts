export * from "./JokeOptions";
export * from "./Dropdown";
