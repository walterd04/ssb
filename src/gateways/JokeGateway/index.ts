export { default as jokeGateway } from "./JokeGateway";
