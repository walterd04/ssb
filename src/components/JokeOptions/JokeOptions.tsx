import React, { useState } from "react";
import styled from "styled-components";

import { jokeGateway } from "../../gateways";
import { Joke, JokeCategory } from "../../types";

import { Dropdown } from "./Dropdown";

interface JokeOptionsProps {
  setJokes: React.Dispatch<React.SetStateAction<Joke[]>>;
  setShowGraphs: React.Dispatch<React.SetStateAction<boolean>>;
  showGraphs: boolean;
}

const JokeOptionsContainer = styled.form`
  display: flex;
  flex-direction: column;
  width: 90%;
  box-shadow: 10px 5px 5px 10px #a6acbe;
  border-radius: 10px;
  background: #ffffff;
  margin: 40px 0;
  padding: 70px 20px;
`;

const Header = styled.h1`
  font-size: 24px;
  font-weight: bold;
  justify-self: center;
  color: #000000;
  text-decoration: underline;
`;

const FormFieldContainer = styled.div`
  display: flex;
  flex-flow: row wrap;
  width: 80%;
  justify-content: space-around;
  align-self: center;
`;

const FormField = styled.div`
  display: flex;
  flex-direction: column;
  & > * {
    margin-right: 10px;
  }
`;

const Label = styled.label`
  font-size: 16px;
  justify-self: center;
`;

const ButtonContainer = styled.div`
  display: flex;
  flex-flow: row wrap;
  align-self: center;
  & > * {
    margin-right: 15px;
  }
`;

const SubmitButton = styled.button`
  padding: 15px;
  border: 1px solid #282c34;
  color: #ffffff;
  background: #282c34;
  border-radius: 10px;
  width: 200px;
  margin: 20px 0;
  align-self: center;
  margin-right: 20px;
`;

export const JokeOptions: React.FC<JokeOptionsProps> = ({
  setJokes,
  setShowGraphs,
  showGraphs,
}: JokeOptionsProps) => {
  const [numJokes, setNumJokes] = useState<number>(10);
  const [category, setCategory] = useState<JokeCategory>("Any");
  const [keywords, setKeywords] = useState<string>("");

  const categoryItems = [
    { label: "Programming", value: "Programming" },
    { label: "Misc", value: "Misc" },
    { label: "Dark", value: "Dark" },
    { label: "Pun", value: "Pun" },
    { label: "Spooky", value: "Spooky" },
    { label: "Christmas", value: "Christmas" },
    { label: "Any", value: "Any" },
  ];

  const submitForm = async (ev: React.FormEvent) => {
    ev.preventDefault();
    const response = await jokeGateway.getJokes(category, keywords, numJokes);
    if (
      response &&
      !response.error &&
      response.jokes &&
      response.jokes.length > 0
    ) {
      setJokes(response.jokes);
    }
  };

  const onNumJokesChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
    setNumJokes((ev.target.value as unknown) as number);
  };

  const onKeywordsChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
    setKeywords(ev.target.value);
  };

  const onShowGraphsClick = (ev: any) => {
    ev.preventDefault();
    setShowGraphs(!showGraphs);
  };

  return (
    <JokeOptionsContainer onSubmit={submitForm}>
      <Header>Welcome to the Random Joke Generator</Header>
      <FormFieldContainer>
        <FormField>
          <Label htmlFor="typeDropdown">Select Type of Joke:</Label>
          <Dropdown
            id="typeDropdown"
            placeholder={category || "Select category"}
            items={categoryItems}
            onChangeEventHandler={(item: any) => setCategory(item.value)}
          />
        </FormField>
        <FormField>
          <Label htmlFor="numJokes">Number of Jokes:</Label>
          <input
            type="number"
            id="numJokes"
            value={numJokes}
            onChange={onNumJokesChange}
          />
        </FormField>
        <FormField>
          <Label>Keyword(s):</Label>
          <input
            type="text"
            id="keywords"
            value={keywords}
            onChange={onKeywordsChange}
          />
        </FormField>
      </FormFieldContainer>
      <ButtonContainer>
        <SubmitButton
          type="submit"
          onSubmit={submitForm}
          data-testid="get-jokes-button"
        >
          Get Jokes
        </SubmitButton>
        <SubmitButton onClick={onShowGraphsClick}>
          Toggle Graphs/Jokes
        </SubmitButton>
      </ButtonContainer>
    </JokeOptionsContainer>
  );
};
