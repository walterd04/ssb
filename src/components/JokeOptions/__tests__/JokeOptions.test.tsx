import React from "react";
import { JokeOptions } from "../JokeOptions";
import { render, waitFor, fireEvent } from "@testing-library/react";
import { jokeGateway } from "../../../gateways";

jest.mock("../../../gateways/JokeGateway");

describe("UI tests", () => {
  it("Renders without crashing, matches snapshot", () => {
    const { baseElement } = render(
      <JokeOptions
        setJokes={jest.fn()}
        setShowGraphs={jest.fn()}
        showGraphs={false}
      />
    );

    expect(baseElement).toMatchSnapshot();
  });
});

describe("Functionality tests", () => {
  it("should call gateway when button is clicked", async () => {
    const { getByTestId } = render(
      <JokeOptions
        setJokes={jest.fn()}
        setShowGraphs={jest.fn()}
        showGraphs={false}
      />
    );

    const getButton = getByTestId("get-jokes-button");
    fireEvent.click(getButton);
    await waitFor(() => {
      expect(jokeGateway.getJokes).toBeCalled();
    });
  });
});
