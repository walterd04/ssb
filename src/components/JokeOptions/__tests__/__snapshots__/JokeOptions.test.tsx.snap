// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI tests Renders without crashing, matches snapshot 1`] = `
<body>
  <div>
    <form
      class="sc-iCoGMd kjJNYD"
    >
      <h1
        class="sc-fujyAs rtCbP"
      >
        Welcome to the Random Joke Generator
      </h1>
      <div
        class="sc-pNWdM fVBpuy"
      >
        <div
          class="sc-jrsJWt hUOEHJ"
        >
          <label
            class="sc-kEqXSa krhUiH"
            for="typeDropdown"
          >
            Select Type of Joke:
          </label>
          <div>
            <div
              class="sc-bdnxRM cPpaHS"
            >
              <select
                data-testid="hidden-select"
                id="typeDropdown"
                name="typeDropdown"
              >
                <option
                  value="Programming"
                >
                  Programming
                </option>
                <option
                  value="Misc"
                >
                  Misc
                </option>
                <option
                  value="Dark"
                >
                  Dark
                </option>
                <option
                  value="Pun"
                >
                  Pun
                </option>
                <option
                  value="Spooky"
                >
                  Spooky
                </option>
                <option
                  value="Christmas"
                >
                  Christmas
                </option>
                <option
                  value="Any"
                >
                  Any
                </option>
              </select>
            </div>
            <div
              class="sc-gtsrHT fQehCQ"
            >
              <div
                class="sc-dlnjwi bjIcPs"
                data-testid="visual-select"
                tabindex="0"
                title="Any"
              >
                <div
                  class="sc-eCApnc fvAlHL"
                >
                  Any
                  <img
                    alt="arrow down"
                    class="sc-gKAaRy iNTfcQ"
                    src="caret.svg"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-jrsJWt hUOEHJ"
        >
          <label
            class="sc-kEqXSa krhUiH"
            for="numJokes"
          >
            Number of Jokes:
          </label>
          <input
            id="numJokes"
            type="number"
            value="10"
          />
        </div>
        <div
          class="sc-jrsJWt hUOEHJ"
        >
          <label
            class="sc-kEqXSa krhUiH"
          >
            Keyword(s):
          </label>
          <input
            id="keywords"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="sc-iqAclL ZboMr"
      >
        <button
          class="sc-crzoAE fwNbVa"
          data-testid="get-jokes-button"
          type="submit"
        >
          Get Jokes
        </button>
        <button
          class="sc-crzoAE fwNbVa"
        >
          Toggle Graphs/Jokes
        </button>
      </div>
    </form>
  </div>
</body>
`;
